simplify code by not using ManagedBlocker.

jbgi opened the pull request:
by sharing ForkJoinTasks through the c.map we can get the same benefits without the complexity.
(This patch is what I meant by the question I asked you this morning).

I think there is no performance loss with this patch.

kabutz commented:
Ah right, thanks for explaining.  On my machine I get a significant performance loss with this approach.  Instead of 8.5s, it completed in 11s.  You could use your approach, but you'd still need to use the ManagedBlocker.

Here are a few thread dumps whilst it is running:

`"ForkJoinPool.commonPool-worker-21" #39 daemon prio=5 os_prio=31 tid=0x00007fc2cc08b000 nid=0x9103 in Object.wait() [0x00007000033af000]
"ForkJoinPool.commonPool-worker-18" #27 daemon prio=5 os_prio=31 tid=0x00007fc2ca0b5000 nid=0x7903 in Object.wait() [0x000070000278b000]
"ForkJoinPool.commonPool-worker-10" #20 daemon prio=5 os_prio=31 tid=0x00007fc2cd82b800 nid=0x6b03 runnable [0x0000700002076000]
"ForkJoinPool.commonPool-worker-7" #17 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7800 nid=0x6503 in Object.wait() [0x0000700001d6d000]
"ForkJoinPool.commonPool-worker-5" #15 daemon prio=5 os_prio=31 tid=0x00007fc2cd829800 nid=0x6103 runnable [0x0000700001b66000]
"ForkJoinPool.commonPool-worker-4" #14 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7000 nid=0x5f03 runnable [0x0000700001a63000]
"ForkJoinPool.commonPool-worker-3" #13 daemon prio=5 os_prio=31 tid=0x00007fc2cd829000 nid=0x5d03 runnable [0x0000700001960000]
"ForkJoinPool.commonPool-worker-2" #12 daemon prio=5 os_prio=31 tid=0x00007fc2cc07a000 nid=0x5b03 in Object.wait() [0x000070000185e000]
"ForkJoinPool.commonPool-worker-1" #11 daemon prio=5 os_prio=31 tid=0x00007fc2cd865800 nid=0x5903 in Object.wait() [0x000070000175b000]

"ForkJoinPool.commonPool-worker-21" #39 daemon prio=5 os_prio=31 tid=0x00007fc2cc08b000 nid=0x9103 in Object.wait() [0x00007000033af000]
"ForkJoinPool.commonPool-worker-18" #27 daemon prio=5 os_prio=31 tid=0x00007fc2ca0b5000 nid=0x7903 runnable [0x000070000278a000]
"ForkJoinPool.commonPool-worker-10" #20 daemon prio=5 os_prio=31 tid=0x00007fc2cd82b800 nid=0x6b03 runnable [0x0000700002075000]
"ForkJoinPool.commonPool-worker-7" #17 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7800 nid=0x6503 in Object.wait() [0x0000700001d6d000]
"ForkJoinPool.commonPool-worker-5" #15 daemon prio=5 os_prio=31 tid=0x00007fc2cd829800 nid=0x6103 runnable [0x0000700001b67000]
"ForkJoinPool.commonPool-worker-4" #14 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7000 nid=0x5f03 runnable [0x0000700001a63000]
"ForkJoinPool.commonPool-worker-3" #13 daemon prio=5 os_prio=31 tid=0x00007fc2cd829000 nid=0x5d03 runnable [0x0000700001961000]
"ForkJoinPool.commonPool-worker-2" #12 daemon prio=5 os_prio=31 tid=0x00007fc2cc07a000 nid=0x5b03 in Object.wait() [0x000070000185e000]
"ForkJoinPool.commonPool-worker-1" #11 daemon prio=5 os_prio=31 tid=0x00007fc2cd865800 nid=0x5903 in Object.wait() [0x000070000175b000]

"ForkJoinPool.commonPool-worker-21" #39 daemon prio=5 os_prio=31 tid=0x00007fc2cc08b000 nid=0x9103 in Object.wait() [0x00007000033af000]
"ForkJoinPool.commonPool-worker-18" #27 daemon prio=5 os_prio=31 tid=0x00007fc2ca0b5000 nid=0x7903 runnable [0x000070000278b000]
"ForkJoinPool.commonPool-worker-10" #20 daemon prio=5 os_prio=31 tid=0x00007fc2cd82b800 nid=0x6b03 runnable [0x0000700002075000]
"ForkJoinPool.commonPool-worker-7" #17 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7800 nid=0x6503 in Object.wait() [0x0000700001d6d000]
"ForkJoinPool.commonPool-worker-5" #15 daemon prio=5 os_prio=31 tid=0x00007fc2cd829800 nid=0x6103 runnable [0x0000700001b67000]
"ForkJoinPool.commonPool-worker-4" #14 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7000 nid=0x5f03 runnable [0x0000700001a64000]
"ForkJoinPool.commonPool-worker-3" #13 daemon prio=5 os_prio=31 tid=0x00007fc2cd829000 nid=0x5d03 in Object.wait() [0x0000700001961000]
"ForkJoinPool.commonPool-worker-2" #12 daemon prio=5 os_prio=31 tid=0x00007fc2cc07a000 nid=0x5b03 runnable [0x000070000185d000]
"ForkJoinPool.commonPool-worker-1" #11 daemon prio=5 os_prio=31 tid=0x00007fc2cd865800 nid=0x5903 runnable [0x000070000175b000]

"ForkJoinPool.commonPool-worker-21" #39 daemon prio=5 os_prio=31 tid=0x00007fc2cc08b000 nid=0x9103 runnable [0x00007000033ae000]
"ForkJoinPool.commonPool-worker-18" #27 daemon prio=5 os_prio=31 tid=0x00007fc2ca0b5000 nid=0x7903 runnable [0x000070000278a000]
"ForkJoinPool.commonPool-worker-10" #20 daemon prio=5 os_prio=31 tid=0x00007fc2cd82b800 nid=0x6b03 runnable [0x0000700002075000]
"ForkJoinPool.commonPool-worker-7" #17 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7800 nid=0x6503 runnable [0x0000700001d6b000]
"ForkJoinPool.commonPool-worker-5" #15 daemon prio=5 os_prio=31 tid=0x00007fc2cd829800 nid=0x6103 runnable [0x0000700001b66000]
"ForkJoinPool.commonPool-worker-4" #14 daemon prio=5 os_prio=31 tid=0x00007fc2ca0a7000 nid=0x5f03 runnable [0x0000700001a63000]
"ForkJoinPool.commonPool-worker-3" #13 daemon prio=5 os_prio=31 tid=0x00007fc2cd829000 nid=0x5d03 runnable [0x0000700001960000]
"ForkJoinPool.commonPool-worker-2" #12 daemon prio=5 os_prio=31 tid=0x00007fc2cc07a000 nid=0x5b03 runnable [0x000070000185e000]
"ForkJoinPool.commonPool-worker-1" #11 daemon prio=5 os_prio=31 tid=0x00007fc2cd865800 nid=0x5903 in Object.wait() [0x000070000175b000]
`

kabutz commented:
Ah wait, you were calling System.currentTimeMillis() every time ... :-)

kabutz commented:
Even with the System.currentTimeMillis() removed, it is still quite a bit slower.

jbgi committed:
simplify code by not using ManagedBlocker.

From my understanding ManagedBlocker is useful when witing on resources not managed by the pool.
When waiting on join, job stealing do occur so using a ManagedBlocker is not necessary.
However is the ManagedBlocker can have the side effect of increasing throughput due to additional threads.
But this is only because the default common pool has size (#processor - 1)
(also maybe. by reducing the probability that a thread could is waiting without having another job to steal).

By using a custom pool with #processor threads or a bit more, one can restore the 100% usage without ManagedBlocker.

jbgi commented:
I did some slight changes with the last commit that should improve on my initial patch:
- use a custom pool of size `#available processors`.
- run main thread on fork join pool so that job stealing can also occur on main thread.
- pre-size the map to avoid some gc.

(but note that all of those optimizations can be applied to the original demo6 commit as well).

From my understanding, ManagedBlocker is useful when waiting on resources not managed by the FJ pool (like a lock or queue).

When waiting on FJTask (join), job stealing do occur so using a ManagedBlocker is not necessary.
However the ManagedBlocker can have the side effect of increasing throughput due to additional threads.
But I think this is mainly because the default common pool has size (#processor - 1).
